packages/runtime/container-runtime/src/blobManager.ts

```diff
@@ -89,10 +92,22 @@ export class BlobManager {
         return this.blobIds.has(id) || this.detachedBlobIds.has(id);
     }
 
+    /**
+     * For a blobId, returns its path in GC's graph. The node path is of the format `/<BlobManager.basePath>/<blobId>`
+     * This path must match the path of the blob handle returned by the createBlob API because blobs are marked
+     * referenced by storing these handles in a referenced DDS.
+     */
+    private getBlobGCNodePath(blobId: string) {
+        return `/${BlobManager.basePath}/${blobId}`;
+    }
+
     public async getBlob(blobId: string): Promise<IFluidHandle<ArrayBufferLike>> {
         const storageId = this.redirectTable?.get(blobId) ?? blobId;
         assert(this.hasBlob(storageId), 0x11f /* "requesting unknown blobs" */);
 
+        // When this blob is retrieved, let the container runtime know.
+        this.blobRetrievedCallback(this.getBlobGCNodePath(blobId));
```

Comment:
The inactive log logic isn't part of the code GC logic so I can see the merits of moving it out of the GarbageCollection class. However, it should not become a generic abstraction that every layer needs to need or care about. It is still a very specific implementation that GC has.
For example, blob manager should not care about logging this event or even know that such an event is being logged. All it should care about is core GC functionality (returning GC data and handling the results from GC). In this particular case, all it does is tells it creator (container runtime) that a blob was requested and container runtime is free to do what it wants with this information.

Reply:
In the `BlobManager` itself, we are already passing in an `IContainerRuntime`. Would it be better to pass in the `ContainerRuntime` itself, or the `IGarbageCollectionRuntime`? I think the `BlobManager` should care about logging this event because it is already attached to our current GC implementation with the written `getGCData` call.

Reply:
Let's do try moving out the inactive log logic if we can. It does not need to be super generic.

Reply:
There are couple of topics here:
1. Passing the whole IContainerRuntime to BlobManager itself is not a good pattern and we should get rid of it. It makes unit testing very difficult as tests have to implement a mock IContaineRuntime which is huge. Also, BlobManager uses very few methods from the runtime and those should either become callbacks or we should pass a smaller sub-interface of IContainerRuntime to it. Similarly, IGarbageCollectionRuntime should not be passed to it as it will only increase the surface area of what BlobManager needs to know about. We should keep this surface area as small as possible.
See https://en.wikipedia.org/wiki/Law_of_Demeter which talks about this principle pretty well.
2. Regarding the logging, BlobManager knows about getGCData because it is core GC functionality and since BlobManager participates in GC, it has to know about it. It would become part of the GC interface that we discussed in the other comment. The inactive logs are a very specific implementation that is temporarily added to find existing bugs in apps. It is not core GC functionality and so other layers should not be concerned about it (as much as possible).
Moreover, the blob manager does not have the GC state to determine when to log this event and what properties to log. All it can do is tell GC that someone requested a blob, please log if necessary. Essentially, the current callback would look something like `logInactiveEventIfBlobInactive`.

Reply:
For moving the inactive log logic out, we need to evaluate how much work it needs and if its worth. I will create a task for it and it can be done separately as it will be a much bigger change.

Reply:
AB#266

Reply:
If the `BlobManager` is our internal implementation, we should be passing in the `GarbageCollection` class into it, and then call something like `this.garbageCollection.nodeUpdated()` - which by the way should actually be named `logIfInactive` and the `private logIfInactive` function should be `_logifInactive` or something like that. Can we change the task to something bigger, as we would want to overhaul the GC class and fix it as much as possible. We should not be creating the Context inside the GC class, it should be passed in from the `ContainerRuntime`.

Reply:
The whole point of having separate classes for managing different functionality is encapsulating the functionality into those classes. ContainerRuntime's functionality is divided in this way. For example, `DataStores` has all functionalities related to data stores, `GarbageCollection` has all functionalities related to GC, `BlobManager` has functionalities related to attachment blobs, `SummaryManager` has functionalities related to summaries, etc. We should not be passing objects of these classes into each other as that defeats the whole purpose. If they do interact, they do so via container runtime by passing a minimal set of properties to it.

I also do not think the GC class needs overhaul. There is some refactor that can be done for sure but the whole idea is for all GC related functionality to be in one place and not leak it all over. There is some leak happening today because of all the staging work but we will stop those.
What you are suggesting is re-implementing / overhauling container runtime itself as this is not just related to GC. That is a very big change and I would suggested starting by creating an issue for doing this and putting in your thoughts into it.

Reply:
I'm saying based on this: https://youtu.be/RlfLCWKxHJ0?t=240, we should just move the creation to the constructor of the container runtime, as that is a simple change, and then eventually fix the container runtime's job of creating these classes by moving it up to the `ContainerRuntime.load` function.

Here's a Draft PR of what I mean: https://github.com/microsoft/FluidFramework/pull/10302

Reply:
This change would enable us to create the GC logger that we actually would want to have to pass as a class down to the `BlobManager`. When we no longer need the class, we would delete the file, and it would be easy to trace back to where it was injected and remove all the necessary dependencies.

Reply:
Callbacks IMO are the hardest functions to trace. It's easier to trace back to an exported function or a class method than a callback.

Reply:
> I'm saying based on this: https://youtu.be/RlfLCWKxHJ0?t=240, we should just move the creation to the constructor of the container runtime, as that is a simple change, and then eventually fix the container runtime's job of creating these classes by moving it up to the `ContainerRuntime.load` function.
> 

I did not get what you mean by this. Move the creation of what to the constructor? Anyway, like I said, this may be a good path forward but its a big change and you should create an issue for it and explain your concern / proposal. That way, others in the team have visibility and can provide feedback.

> This change would enable us to create the GC logger that we actually would want to have to pass as a class down to the BlobManager. When we no longer need the class, we would delete the file, and it would be easy to trace back to where it was injected and remove all the necessary dependencies.

I am not sure I see how this will work from your draft. I believe what you mean is that the logic that determines whether an object is inactive or not will have to move out of the GC class which may be fine. Please feel free to prototype this and / or put your ideas into the issue I created for this - [AB#266](https://dev.azure.com/fluidframework/internal/_workitems/edit/266)